ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp

```diff
@@ -17,37 +17,63 @@ size_t round_up(size_t value, size_t multiple) {
     return ((value + multiple - 1) / multiple) * multiple;
 };
 
-Alignment legacyShapeToAlignment(const ttnn::Shape& shape) {
-    auto legacy_padded_shape = shape.padded_shape();
-    if (shape.logical_shape() == legacy_padded_shape) {
+Alignment legacyShapeToAlignment(const ttnn::Shape& shape, const PageConfig& page_config, const MemoryConfig& memory_config) {
+    const auto& logical_shape = shape.logical_shape();
+    const auto& legacy_padded_shape = shape.padded_shape();
+    if (logical_shape == legacy_padded_shape) {
         return Alignment{};
     }
 
     const auto rank = legacy_padded_shape.rank();
-    ttnn::SmallVector<uint32_t> values(rank);
-
-    if(rank >= 1) {
-        values[rank - 1] = legacy_padded_shape[rank - 1];
-    }
-    if(rank >= 2) {
-        values[rank - 2] = legacy_padded_shape[rank - 2];
-    }
+    bool alignment_can_be_2D = true;
     for (int i = rank - 3; i >= 0; i--) {
-        values[i] = legacy_padded_shape[i] * values[i + 1];
+        alignment_can_be_2D &= logical_shape[i] == legacy_padded_shape[i];
     }
 
-    Alignment result(std::move(values));
-    return result;
+    if (memory_config.shard_spec.has_value()) {
+        TT_FATAL(alignment_can_be_2D, "Tensor with shape {} cannot be sharded because alignment will have rank greater than 2!", shape);
+        const auto& shard_spec = memory_config.shard_spec.value();
+        TT_FATAL(shard_spec.mode == ShardMode::PHYSICAL, "Shard mode {} has to be ShardMode::PHYSICAL for legacyShapeToAlignment!", shard_spec.mode);
+        if (page_config.is_row_major()) {
+            return Alignment{shard_spec.shape[1]};
+        }
+        return Alignment{};
+    } else {
+        if (alignment_can_be_2D) {
```

Comment:
@TT-BrianLiu Do we need this check? The code inside seems very similar, maybe we can unify it more?